tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,250 @@
+    std::vector<int64_t> devices(num_devices_);
+    std::iota(devices.begin(), devices.end(), 0);
+    world_communicator_ =
+        communicator->getBackendForTeam(devices, params.backend_type);
+
+    // Define I/O and intermediate Tensor shapes
+    std::vector<int64_t> ta_unsharded_sizes = {
+        params.S,
+        num_devices_,
+        params.M / params.S,
+        params.K / num_devices_,
+        1};
+    std::vector<int64_t> ta_sizes = {
+        params.S, 1, params.M / params.S, params.K / num_devices_, 1};
+    std::vector<int64_t> tb_unsharded_sizes = {
+        1, num_devices_, 1, params.K / num_devices_, params.N};
+    std::vector<int64_t> tb_sizes = {
+        1, 1, 1, params.K / num_devices_, params.N};
+    std::vector<int64_t> tc_locally_reduced_sizes = {
+        params.S, 1, params.M / params.S, params.N};
+    std::vector<int64_t> tc_sizes = {
+        params.S, 1, params.M / (params.S * num_devices_), params.N};
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    at::TensorOptions options =
+        at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+    auto ta_unsharded = at::randn(ta_unsharded_sizes, options);
+    auto tb_unsharded = at::randn(tb_unsharded_sizes, options);
+    ta_ = at::empty(ta_sizes, options);
+    ta_.copy_(getSlice(ta_unsharded, 1, my_device_index_));
+    tb_ = at::empty(tb_sizes, options);
+    tb_.copy_(getSlice(tb_unsharded, 1, my_device_index_));
+
+    // We pre-allocate the output and some intermediate buffers so we do not
+    // rely on torch allocator, which do not behave well with multi-stream
+    // programming.
+    tc_locally_reduced_ = at::empty(tc_locally_reduced_sizes, options);
+    tc_ = at::empty(tc_sizes, options);
+
+    // compute the expected output for data correctness validation
+    auto tc_unsharded_unreduced = ta_unsharded * tb_unsharded;
+    auto tc_unsharded_expected = at::sum(tc_unsharded_unreduced, {1, 3});
+    auto tc_unsharded_expected_reshaped = at::reshape(
+        tc_unsharded_expected,
+        {params.S,
+         num_devices_,
+         params.M / (params.S * num_devices_),
+         params.N});
+    tc_expected_ =
+        getSlice(tc_unsharded_expected_reshaped, 1, my_device_index_);
+
+    // Debug print
+    if (communicator->deviceId() == 0 && debug_print) {
+      std::cout << params << std::endl;
+      std::cout << "ta_.sizes()=" << ta_.sizes() << std::endl;
+      std::cout << "tb_.sizes()=" << tb_.sizes() << std::endl;
+      std::cout << "tc_locally_reduced_.sizes()=" << tc_locally_reduced_.sizes()
+                << std::endl;
+      std::cout << "tc_.sizes()=" << tc_.sizes() << std::endl;
```

Comment:
Debug prints in general go to std::cerr.

Reply:
ok, I change to using `debug()`